Simulation and Modeling Lab  
**Course Code:** CSE 413  
 **Course Teacher:** Audity Ghosh, Lecturer, Department of Computer Science and Engineering, University of Information Technology and Sciences (UITS)  
**Date:** 07 Aug, 2026  


Unique Matrix Value Investigation

In [18]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns

""" 📝 Task Description
 1. Create Your Personal Matrix
 Let:
  My Student ID = 0432320005101071
 ● d1 = last digit of your student = 1
 ● d2 = second last digit of your student ID = 7

#Create the matrix:
   A = [[d1+2, d2+1],
        [2d1,  d2+2]
        ]
 """
d1 = 1
d2 = 7
A = np.array([[d1+2, d2+1],
              [2*d1, d2+2]]
             )
print(A)

# 2. Analyze Matrix A
"""
Using NumPy, compute:
● Shape of A
● Determinant
● Rank
● Eigenvalues
● Inverse (only if determinant ≠ 0)
"""
shape = A.shape
print("Shape of A:", shape)

det = np.linalg.det(A)
print("Determinant of A:", det)

rank = np.linalg.matrix_rank(A)
print("Rank of A:", rank)

eigenvalues, eigenvectors = np.linalg.eig(A)
print("Eigenvalues of A:", eigenvalues)
print("Eigenvectors of A:\n", eigenvectors)
# determinant is ≠ 0
#Determinant of A = 11.000000000000002
inverse = np.linalg.inv(A)
print("Inverse of A:\n", inverse)



[[3 8]
 [2 9]]
Shape of A: (2, 2)
Determinant of A: 11.000000000000002
Rank of A: 2
Eigenvalues of A: [ 1. 11.]
Eigenvectors of A:
 [[-0.9701425  -0.70710678]
 [ 0.24253563 -0.70710678]]
Inverse of A:
 [[ 0.81818182 -0.72727273]
 [-0.18181818  0.27272727]]


3. Value Change Experiment

In [19]:
"""
Create a new matrix B by changing exactly one value in A:
● Add +1 to one element of your choice
● You must clearly state which element you changed
"""
B = A.copy()
B[0, 0] += 1
print("Matrix B:\n", B)

Matrix B:
 [[4 8]
 [2 9]]


4. Re-analyze Matrix B

In [20]:
#Compute the same properties as in Step 2.
"""
Using NumPy, compute:
● Shape of B
● Determinant
● Rank
● Eigenvalues
● Inverse (only if determinant ≠ 0)
"""
shape = B.shape
print("Shape of B:", shape)

det = np.linalg.det(B)
print("Determinant of B:", det)

rank = np.linalg.matrix_rank(B)
print("Rank of B:", rank)

eigenvalues, eigenvectors = np.linalg.eig(B)
print("Eigenvalues of B:", eigenvalues)
print("Eigenvectors of B:\n", eigenvectors)
# determinant is ≠ 0
# Determinant of B = 19.999999999999996
inverse = np.linalg.inv(B)
print("Inverse of B:\n", inverse)


Shape of B: (2, 2)
Determinant of B: 19.999999999999996
Rank of B: 2
Eigenvalues of B: [ 1.78300943 11.21699057]
Eigenvectors of B:
 [[-0.96368025 -0.74250951]
 [ 0.26705875 -0.66983552]]
Inverse of B:
 [[ 0.45 -0.4 ]
 [-0.1   0.2 ]]


5. Comparison & Explanation

**1.   How did the determinant change and why?**

**Answer:** The determinant increased from 11 to 20 (a net change of +9).
Explanation:
The Matrix Determinant Lemma (Rank-One Update Theory)
In advanced linear algebra, changing a single element of a matrix is classified as a rank-one update (adding a matrix of rank 1 to $A$). The Matrix Determinant Lemma formalizes how the determinant responds when a matrix undergoes a rank-one modification ($A + u v^T$).

For a 2x2 matrix,
   det = ad - bc.
   Modifying a11 from a to a+1
   the det(B) = (a+1)d - bc = (ad - bc) + d = det(A) + d. Therefore, adding +1 to a11 increased the determinant by the value of a22 (9).  
2.  Did the rank change? Explain.

**Answer:**The rank is not changed beacuse A square matrix maintains full rank as long as its determinant is non-zero (det ≠ 0). Because det(B) = 20  ≠ 0, the rows [4, 8] and [2, 9] remain linearly independent and span the full two-dimensional space ($\mathbb{R}^2$).


3.   How did the eigenvalues respond to the value change?

**Answer:**eigenvalues shifted upward, changing from [1.0, 11.0] in A to approximately [1.78, 11.22] in B.
Eigenvalues are tied directly to the matrix's diagonal sum and determinant. Because both of those total values increased when we changed 3 to 4, the eigenvalues had to grow to match them.


4.   Is B easier or harder to invert than A? Why?

**Answer:**Analytically, they are equally simple; numerically, B is slightly easier (more stable).

Explanation: Computationally, both A and B require identical steps using the 2x2 adjugate inverse formula: B^-1 = (1 / det(B)) * adj(B). However, because det(B) = 20 is larger and further away from zero than det(A) = 11, matrix B has a better (lower) condition number, making it less sensitive to floating-point rounding errors during numerical inversion





